### PR TITLE
[1.7] Fix WorkloadEntry Updates

### PR DESCRIPTION
**Pull request by howardjohn**
Fixes https://github.com/istio/istio/issues/27304

WIP. This is a ton of changes and mostly manual cherrypick.

### CODE REVIEW

**Comment by howardjohn** [comment]
@ramaraochavali @hzxuzhonghu can you give this another review? The code in master is substantially different so it is a pretty invasive cherrypick

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 1]
```diff
@@ -342,6 +344,23 @@ func (c *Controller) onServiceEvent(curr interface{}, event model.Event) error {
...
 	}
 
+	// We also need to update when the Service changes. For Kubernetes, a service change will result in Endpoint updates,
+	// but workload entries will also need to be updated.
+	if event == model.EventAdd || event == model.EventUpdate {
+		// Build IstioEndpoints
+		endpoints := c.endpoints.buildIstioEndpointsWithService(svc.Name, svc.Namespace, svcConv.Hostname)
+		var fep []*model.IstioEndpoint
+		if features.EnableK8SServiceSelectWorkloadEntries {
+			fep = c.collectWorkloadInstanceEndpoints(svcConv)
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:354]: nit: we can only run L351 when `fep` not nil

File: pilot/pkg/serviceregistry/serviceentry/servicediscovery.go [hunk 2]
```diff
@@ -235,66 +238,73 @@ func (s *ServiceEntryStore) serviceEntryHandler(old, curr model.Config, event mo
...
 	// Used to indicate if this event was fired for a pod->workloadentry conversion
 	// and that the event can be ignored due to no relevant change in the workloadentry
 	redundantEventForPod := false
 
+	var addressToDelete string
+
 	s.storeMutex.Lock()
 	// this is from a pod. Store it in separate map so that
 	// the refreshIndexes function can use these as well as the store ones.
+	k := si.Name + "~" + si.Namespace
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/serviceentry/servicediscovery.go:291]: hmm. this should be consistent with `namespace/name`, which is widely used

**Review by hzxuzhonghu — APPROVED** [review]

**Review by jwendell — APPROVED** [review]

**Pull request merged**